packages/daemon/src/daemon.js

```diff
@@ -927,22 +934,31 @@ const makeDaemonCore = async (
   };
 
   /**
-   * Formulates a `handle` formula and synchronously adds it to the formula graph.
-   * The returned promise is resolved after the formula is persisted.
+   * Unlike other formulate functions, formulateNumberedHandle *only* writes a
+   * formula to the formula graph and does not attempt to incarnate it.
+   * This is to break an incarnation cycle between agents and their handles.
+   * The agent must be incarnated first, contains its own handle object, and
+   * produces a agentIdForHandle entry as a side-effect.
+   * Explicitly incarnating the handle formula later simply looks up the handle
+   * reference on the already-incarnated agent.
    *
    * @param {string} formulaNumber - The formula number of the handle to formulate.
-   * @param {string} targetId - The formula identifier of the handle's target.
-   * @returns {import('./types.js').FormulateResult<import('./types.js').ExternalHandle>} The formulated handle.
+   * @param {string} agentId - The formula identifier of the handle's agent.
+   * @returns {Promise<string>}
    */
-  const formulateNumberedHandle = (formulaNumber, targetId) => {
+  const formulateNumberedHandle = async (formulaNumber, agentId) => {
     /** @type {import('./types.js').HandleFormula} */
     const formula = {
       type: 'handle',
-      target: targetId,
+      agent: agentId,
     };
-    return /** @type {import('./types').FormulateResult<import('./types').ExternalHandle>} */ (
-      formulate(formulaNumber, formula)
-    );
+    await persistencePowers.writeFormula(formulaNumber, formula);
+    const id = formatId({
+      number: formulaNumber,
+      node: ownNodeIdentifier,
+    });
+    formulaForId.set(id, formula);
```

Comment:
We should perhaps extract the first ~10 lines of `formulate()` into a new function and replace this inlined formula graph mutation with that.

I don't believe it matters in our usage of `formulateNumberedHandle()`, but `formulate()` also does not call `writeFormula()` until after mutating `formulaForId`.

Reply:
This is interesting and I’d like to revisit it. There might be a case for awaiting on waitFormula before this. It might help to test writeFormula failures and recovery from those failures on restart.

Reply:
https://github.com/endojs/endo/issues/2207